File: Bamilo/DataManager.m
 //
//  DataManager.m
//  Bamilo
//
//  Created by Narbeh Mirzaei on 2/5/17.
//  Copyright © 2017 Rocket Internet. All rights reserved.
//

#import "DataManager.h"
#import "RICart.h"
#import "RIForm.h"
#import "RICategory.h"
#import "OrderList.h"

@implementation DataManager

static DataManager *instance;

- (instancetype)init {
    if (self = [super init]) {
        self.requestManager = [[RequestManager alloc] initWithBaseUrl:[NSString stringWithFormat:@"%@%@", [RIApi getCountryUrlInUse], RI_API_VERSION]];
    }
    
    return self;
}

+ (instancetype)sharedInstance {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [DataManager new];
    });
    
    return instance;
}

#pragma mark - Public Methods
- (void) processResponse:(RIApiResponse)response ofClass:(Class)aClass forData:(ResponseData *)data errorMessages:(NSArray *)errorMessages completion:(DataCompletion)completion {
    if(completion) {
        if(response == RIApiResponseSuccess && data) {
            NSMutableDictionary *payload = [NSMutableDictionary dictionary];
            
            DataMessageList *dataMessages = data.messages;
            if(dataMessages) {
                [payload setObject:dataMessages forKey:kDataMessages];
            }

            if([aClass conformsToProtocol:@protocol(JSONVerboseModel)]) {
                [RICountry getCountryConfigurationWithSuccessBlock:^(RICountryConfiguration *configuration) {
                    [payload setObject:[aClass parseToDataModelWithObjects:@[ ((ResponseData *)data).metadata, configuration ]] forKey:kDataContent];
                    [self handlePayload:payload completion:completion];
                } andFailureBlock:^(RIApiResponse apiResponse, NSArray *errorMessages) {
                    completion(nil, [self getErrorFrom:apiResponse errorMessages:errorMessages]);
                }];
            } else {
                id metadata = data.metadata;
                if(metadata && aClass) {
                    NSError *error;
                    id dataModel = [[aClass alloc] init];
                    [dataModel mergeFromDictionary:data.metadata useKeyMapping:YES error:&error];
                    
                    if(error == nil) {
                        [payload setObject:dataModel forKey:kDataContent];
                    }
                }
                
                [self handlePayload:payload completion:completion];
            }
        } else {
            completion(nil, [self getErrorFrom:response errorMessages:errorMessages]);
        }
    }
}

- (NSError *)getErrorFrom:(RIApiResponse)response errorMessages:(NSArray *)errorMessages {
    return [NSError errorWithDomain:@"com.bamilo.app.ios" code:response userInfo:(errorMessages ? @{ kErrorMessages: errorMessages } : nil)];
}

-(void) handlePayload:(NSMutableDictionary *)payload completion:(DataCompletion)completion {
    if(payload.allKeys.count > 1) {
        completion(payload, nil);
    } else if([payload objectForKey:kDataContent]) {
        completion([payload objectForKey:kDataContent], nil);
    } else {
        completion([payload objectForKey:kDataMessages], nil);
    }
}

@end

File: DataManager.swift
//
//  DataManager.swift
//  Bamilo
//
//  Created by Narbeh Mirzaei on 6/3/17.
//  Copyright © 2017 Rocket Internet. All rights reserved.
//

import ObjectMapper

typealias DataClosure = (_ data: Any?, _ error: NSError?) -> Void

struct DataManagerKeys {
    static let DataMessages = "DataMessages"
    static let DataContent = "DataContent"
}

class DataManagerSwift {
    static private(set) var requestManager: RequestManagerSwift = {
        if let apiBaseUrl = AppUtility.getInfoConfigs(for: AppKeys.APIBaseUrl) as? String,
            let apiVersion = AppUtility.getInfoConfigs(for: AppKeys.APIVersion) as? String {
            return RequestManagerSwift(with: "\(apiBaseUrl)v\(apiVersion)")
        }
        return RequestManagerSwift(with: nil)
    }()
    
    func processResponse(_ responseType: Int, aClass: Any?, data: ApiResponseData?, errorMessages: [Any]?, completion: DataClosure?) {
        if let completion = completion {
            guard let data = data else {
                return completion(nil, self.createError(responseType, errorMessages: errorMessages))
            }
            var payload = [String: Any]()
            if let messages = data.messages {
                payload[DataManagerKeys.DataMessages] = messages
            }
            if let serviceData = data.metadata as [String:Any]? {
                if let mappableClass = aClass as? Mappable.Type, let responseObj = mappableClass.init(JSON: serviceData) {
                    payload[DataManagerKeys.DataContent] = responseObj
                    self.handlePayload(payload: payload, data: data, completion: completion)
                } else if let mappableClassVerbose = aClass as? JSONVerboseModel.Type {
                    RICountry.getConfigurationWithSuccessBlock({ (configuration) in
                        payload[DataManagerKeys.DataContent] = mappableClassVerbose.parseToDataModel(with: [serviceData, configuration!])
                        self.handlePayload(payload: payload, data: data, completion: completion)
                    }, andFailureBlock: { (apiResponse, errorMessages) in
                        if let errorMessages = errorMessages, let apiResponseType = self.map(statusCode: apiResponse) {
                            completion(nil, self.createError(apiResponseType, errorMessages:errorMessages))
                        }
                    })
                } else if let mappableClass = aClass as? BaseModel.Type {
                    let dataModel = mappableClass.init()
                    try? dataModel.merge(from: serviceData, useKeyMapping: true, error: ())
                    payload[DataManagerKeys.DataContent] = dataModel
                    self.handlePayload(payload: payload, data: data, completion: completion)
                } else { // can not be parse as expected, just pass data
                    self.handlePayload(payload: payload, data: data, completion: completion)
                }
            } else {
                self.handlePayload(payload: payload, data: data, completion: completion)
            }
        }
    }
    
    func createError(_ responseType: Int, errorMessages: [Any]?) -> NSError? {
        guard let errorMessages = errorMessages else {
            return NSError(domain:AppValues.Domain, code:responseType, userInfo:nil)
        }
        return NSError(domain:AppValues.Domain, code:responseType, userInfo:[ AppKeys.ErrorMessages : errorMessages ])
    }
    
    //MARK: Private Methods
    private func handlePayload(payload: [String:Any], data: ApiResponseData?, completion: @escaping DataClosure) {
        if payload.count > 1 {
            completion(payload, nil)
        } else if payload[DataManagerKeys.DataContent] != nil {
            completion(payload[DataManagerKeys.DataContent], nil)
        } else if payload[DataManagerKeys.DataMessages] != nil {
            completion(payload[DataManagerKeys.DataMessages], nil)
        } else {
            completion(data, nil)
        }
    }
    
    private func map(statusCode: RIApiResponse) -> Int? {
        switch statusCode {
            case .success:
                return 200
            case .authorizationError:
                return NSURLErrorUserAuthenticationRequired
            case .timeOut:
                return NSURLErrorTimedOut
            case .badUrl:
                return NSURLErrorBadURL
            case .unknownError:
                return NSURLErrorUnknown
            case .apiError:
                return NSURLErrorBadServerResponse
            case .noInternetConnection:
                return NSURLErrorNotConnectedToInternet
            case .maintenancePage:
                return NSURLErrorCannotConnectToHost
            case .kickoutView:
                return NSURLErrorUnsupportedURL
        }
    }
}
